File: src/main/java/org/voegtle/weatherwidget/notification/WeatherStationCheck.java
package org.voegtle.weatherwidget.notification;

import org.voegtle.weatherwidget.data.WeatherData;
import org.voegtle.weatherwidget.location.LocationIdentifier;
import org.voegtle.weatherwidget.location.WeatherLocation;
import org.voegtle.weatherwidget.preferences.ApplicationSettings;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

public class WeatherStationCheck {
  private static int THRESHOLD = 20 * 60 * 1000; // 20 min
  private final ApplicationSettings configuration;

  List<WeatherAlert> alerts = new ArrayList<>();

  public WeatherStationCheck(ApplicationSettings configuration) {
    this.configuration = configuration;
  }

  public List<WeatherAlert> checkForOverdueStations(HashMap<LocationIdentifier, WeatherData> data) {
    alerts.clear();
    for (WeatherLocation location : configuration.getLocations()) {
      if (location.getPreferences().isAlertActive()) {
        buildAlert(data.get(location.getKey()));
      }
    }

    return alerts;
  }

  private void buildAlert(WeatherData data) {
    Date now = new Date();
    if ((now.getTime() - data.getTimestamp().getTime()) > THRESHOLD) {
      alerts.add(new WeatherAlert(data.getLocation(), data.getTimestamp()));
    }
  }
}

File: src/main/java/org/voegtle/weatherwidget/notification/WeatherStationCheck.kt
package org.voegtle.weatherwidget.notification

import org.voegtle.weatherwidget.data.WeatherData
import org.voegtle.weatherwidget.location.LocationIdentifier
import org.voegtle.weatherwidget.location.WeatherLocation
import org.voegtle.weatherwidget.preferences.ApplicationSettings

import java.util.ArrayList
import java.util.Date
import java.util.HashMap

class WeatherStationCheck(private val configuration: ApplicationSettings) {

  internal var alerts: MutableList<WeatherAlert> = ArrayList()

  fun checkForOverdueStations(data: HashMap<LocationIdentifier, WeatherData>): List<WeatherAlert> {
    alerts.clear()
    for (location in configuration.locations) {
      if (location.preferences.alertActive) {
        buildAlert(data[location.key])
      }
    }

    return alerts
  }

  private fun buildAlert(data: WeatherData?) {
    if (data != null) {
      val now = Date()
      if (now.time - data.timestamp!!.time > THRESHOLD) {
        alerts.add(WeatherAlert(data.location, data.timestamp!!))
      }
    }
  }

  companion object {
    private val THRESHOLD = 20 * 60 * 1000 // 20 min
  }
}
